design/vgdp-micro-service/vgdp-micro-service.md

```diff
@@ -0,0 +1,186 @@
+```velero vgdp-ms backup --this-pod xxx --this-container xxx --this-volume xxx --resource-timeout xxx --log-format xxx --log-level xxx```  
+Or:  
+```velero vgdp-ms restore --this-pod xxx --this-container xxx --this-volume xxx --resource-timeout xxx --log-format xxx --log-level xxx```  
+
+The first one is for backup and the other one is for restore.  
+Below are the parameters of the commands:  
+**this-pod**: Deliver the backupPod/restorePod's name to VGDP-MS. Even though VGDP-MS is running inside the backupPod/restorePod, it is not easy to get the name of pod it is running in, so it is simply delivered through the parameter.  
+**this-container**: When backupPod/restorePod is started, the cluster may insert other containers to the pod, so we cannot assume that the container running VGDP-MS is always the first container in the pod. So the container name is simply delivered through the parameter.  
+**this-volume**: The name of the snapshot/target volume. VGDP-MS accesses the snapshot/target volume by matching this-volume to the volumeMount/volumeDevice in backupPod/restorePod's spec, so that it could access the snapshot/target volume locally.  
+**resource-timeout**: resource-timeout is used to control the timeout for operations related to resources. It has the same meaning with the resource-timeout for node-agent.  
+**log-format** and **log-level**: This is to control the behavior of log generation inside VGDP-MS. 
+
+In order to have the same capability and permission with node-agent, below pod configurations are inherited from node-agent and set to backupPod/restorePod's spec:
+- Volumes: Some configMaps will be mapped as volumes to node-agent, so we add the same volumes of node-agent to the backupPod/restorePod
+- Environment Variables
+- Security Contexts
+
+### VGDP Watcher
+#### Dual mode event watch
+The primary task of VGDP Watcher is to watch the status change from backupPod/restorePod or the VGDP instance, so as to inform the data mover controller in below situations:
+- backupPod/restorePod starts
+- VGDP instance starts
+- Progress update
+- VGDP instance completes/fails/cancelled
+- backupPod/restorePod stops
+
+We use two mechanism to make the watch:  
+**Pod Phases**: VGDP Watcher watches the backupPod/restorePod's phases updated by Kubernetes. That is, VGDP Watcher creates an informer to watch the pod resource for the backupPod/restorePod and detect that the pod reaches to one of the terminated phases (i.e., PodSucceeded, PodFailed). We also check the availability & status of the backupPod/restorePod at the beginning of the watch so as to detect the starting of the backupPod/restorePod.  
+**Custom Kubernetes Events**: VGDP-MS generates Kubernetes events and associates them to the backupPod/restorePod at the time of VGDP instance starting/stopping and progress update, then VGDP Watcher creates another informer to watch the Event resource associated to the backupPod/restorePod.  
```

Comment:
What's the specific use case for this?  reporting the transfer status?
IMHO, we wanna try to make the code that runs in backupPod less k8s-aware or k8s-dependent, so we can test it or verify certain cases more easily.  But this is an immature thought and we can have further discussions on this topic.

Reply:
`Custom Kubernetes Events` doesn't mean a mechanism out of k8s scope, so it is still k8s aware/dependent.
Here the Events refer to the Kubernetes Event resource, that is, we generate events to the backupPod/restorePod during the data movement.
We use the Events to transfer message to the controller during the data movement, since the `Pod Phase` and `Pod Terminate Message` mechanism can only deliver info at the time of pod terminates.

Reply:
After offline discussion, we agreed to choose k8s events as the approach for backup/restore pod to communicate to the node-agent, but maybe point the `involvedObject` to the DUCR is a better choice.

Reply:
After some further checks, I can see below negative points if we want to make `involvedObject` as DUCR/DDCR:
1. Kubernetes' etcd is not good at to store huge number of objects, this is why the event objects are automatically removed after 1 hour(by default). Since DUCR/DDCR are to be permanently kept, it will increase the burden to etcd when there are lots of DUCR/DDCR. On the other hand, since the events are deleted after 1 hour, we have no way to keep the events for further use, so there is no gain to keep the events with DUCR/DDCR.
2. The ms-service solution introduced in this design is capable to be reused by other requirements in future and Event is a core part of this solution. Adding DUCR/DDCR as `involvedObject` requires adding one more input parameter for the ms-service. As a comparison, originally, we only require `this-pod`, `this-container` and `this-volume` as the input.

Reply:
After a further discussion, we decided to associate the events to DUCR/DDCR

Reply:
Done on the changes